eeschema/sch_edit_frame.cpp

```diff
@@ -1519,7 +1519,7 @@ void SCH_EDIT_FRAME::RecalculateConnections()
 
     // Ensure schematic graph is accurate
     for( const auto& sheet : list )
-        SchematicCleanUp( true, sheet.LastScreen() );
+        SchematicCleanUp( false, sheet.LastScreen() );
```

Comment:
If SchematicCleanup changes anything, these changes will be lost to the undo/redo stack.  If it cannot change anything, we should remove it or clear the undo/redo stack.

Reply:
My thinking here is that there are two types of callers to `SchematicCleanUp`: those that clear the undo/redo stack anyway (like on file open) and those that call it explicitly after doing some modification (like most of the call sites in `bus-wire-junction.cpp`).  This loop is here to make sure that the entire schematic is in a sane state before calculating connectivity, and so I don't want to create an undo commit here because this is an automatic action that isn't part of a normal user action.

In cases where there *was* an undoable action, like for example placing a wire, `SchematicCleanUp` will get called (at least) twice: once in `bus-wire-junction` where the changes are saved in the undo stack, and then a second time here (where it will have no impact).

Does that make sense?  I guess I could make it more bulletproof by forcing a clear of the undo stack if you pass `false` here and moving that out of the call sites that clear the stack anyway.

Reply:
Ah, I see (I think).  This would mean that the correct way of working this is that if we are cleaning up the non-current sheet, we add to the undo stack.  I remain concerned about the lack of append here.  This means that you can never call this function without first adding something to the undo stack.  Otherwise, it will be added to the previous commit.

Reply:
To be more clear, there are three paths for SchematicCleanUp:

First path:

1. User takes some action that modifies wires
2. Action code creates a commit
3. Action code calls `SchematicCleanUp( true )` which appends to the commit
4. Action code calls `OnModify()` which will then call `SchematicCleanUp( false )` in `RecalculateConnections()`

Second path:

1. User takes some action that does not modify wires
2. Action code calls `OnModify()` which will then call `SchematicCleanUp( false )` in `RecalculateConnections()`.  `SchematicCleanUp` takes no action because there was nothing to fix, and so undo is not needed here.

Third path:

1. Schematic is opened
2. `SchematicCleanUp( false )` is called to put the schematic into a consistent state.  Undo is not needed here because the user took no explicit action, we're just "fixing" the schematic that was saved on disc.

So, as far as I can see, there is never a reason why `SchematicCleanUp` should create a new undo commit -- it should always either append to the previous commit, or not save the undo state at all.  Does that make sense? Let me know if I'm missing something.